keps/sig-architecture/3935-oldest-node-newest-control-plane/README.md

```diff
@@ -0,0 +1,799 @@
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+  - sig-cluster-lifecycle: [2023-04-04](https://docs.google.com/document/d/1Gmc7LyCIL_148a9Tft7pdhdee0NBHdOfHS1SAF0duI4/edit#bookmark=id.dlswqqyv7xse)
+  - sig-storage: [2023-04-06](https://docs.google.com/document/d/1-8KEG8AjAgKznS9NFm3qWqkGyCHmvU6HVl0sk5hwoAE/edit#bookmark=id.3pg0lcwvpqce)
+  - sig-architecture: [2023-04-06](https://docs.google.com/document/d/1BlmHq5uPyBUDlppYqAAzslVbAO8hilgjqZUTaNXUhKM/edit#bookmark=id.b6nofrhcysqn)
+  - sig-node: [2023-04-11](https://docs.google.com/document/d/1Ne57gvidMEWXR70OxxnRkYquAoMpt56o75oZtg-OeBg/edit#bookmark=kix.clb5o8ybahss)
+  - sig-network: [2023-04-13](https://docs.google.com/document/d/1_w77-zG_Xj0zYvEMfQZTQ-wPP4kXkpGD8smVtW_qqWM/edit#bookmark=id.ehzitjq2ldlo)
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes testing and expanding the supported skew between core node and control plane components
+by one version from n-2 to n-3, so that node components (kubelet and kube-proxy) for the **oldest** supported minor version
```

Comment:
So we discussed this some on slack but it doesn't seem to have made it into the doc...

I don't think you can really count kube-proxy as being the same sort of thing as kubelet. In particular, kube-proxy may be deployed as a component of a third-party network plugin, and thus be skewed relative to either or both of the control plane and kubelet during upgrades.

If this KEP is supposed to be "everything in k/k should support n-3 skew with everything else in k/k", then you need to consider both kube-proxy/control-plane skew and kube-proxy/kubelet skew, in which case the 1.27 section should mention that KEP-3178 broke kube-proxy 1.27 vs kubelet 1.24. (But 1.28/1.25 is fine.)

If the KEP is just "everything in k/k should support n-3 skew with the control plane but not necessarily with other k/k components" then it may be worth clarifying that in the Non-Goals?

Reply:
Happy to add more details, but based on our slack conversation, I thought the skew KEP-3178 relied on avoiding was a 1.24 kube-proxy with a 1.27 kubelet. I don't *think* that's contemplated here.

There are two kube-proxy deployment mechanisms I'm aware of:
* versioned with node (static pod, etc); the implication of this KEP would mean that 1.24 kube-proxy and kubelet would have needed to work together and work against a 1.27 API server.
* versioned with control-plane (daemonset, etc); the implication of this KEP would mean a 1.24 kubelet and a 1.27 kube-proxy would have needed to work together and work against a 1.27 API server.

Is there a rollout mechanism I'm missing that would result in:
* kube-apiserver is 1.27
* kube-proxy is 1.24
* kubelet is 1.27

Reply:
as an aside, I know some updates to the skew policy for kube-proxy were made after the sig-network discussion, but https://kubernetes.io/releases/version-skew-policy/#kube-proxy is still sort of confusing and indicates kubelet/kube-proxy versions are tightly coupled

Reply:
The third deployment mechanism is "kube-proxy is deployed by the network plugin, which is separate from both the node and the control plane". So a cluster could do
1. upgrade control-plane from 1.24 to 1.27
2. upgrade kubelet from 1.24 to 1.27
3. upgrade network plugin, including upgrading kube-proxy from 1.24 to 1.27

and between steps 2 and 3 they'd have kube-proxy 1.24 running against kubelet 1.27 and kube-apiserver 1.27.

> but https://kubernetes.io/releases/version-skew-policy/#kube-proxy is still sort of confusing and indicates kubelet/kube-proxy versions are tightly coupled

oops, I guess there were originally two references to kube-proxy/kubelet coupling and I only removed one.

Reply:
opened https://github.com/kubernetes/website/pull/40932/files for ^^

Reply:
answering:

> Is there a rollout mechanism I'm missing that would result in:
> 
> * kube-apiserver is 1.27
> * kube-proxy is 1.24
> * kubelet is 1.27

we synced offline and @danwinship indicated there are deployments using daemonsets that are not versioned with nodes OR control planes, so that configuration is possible. I'll update this KEP to mention the KEP-3178 implications for 1.24/1.27 kube-proxy/kubelet skew for anyone deploying kube-proxy that way